src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs

```diff
@@ -130,95 +130,123 @@ public static SafeFreeCredentials AcquireCredentialsHandle(X509Certificate certi
             return AcquireCredentialsHandle(direction, secureCredential);
         }
 
-        public static SecurityStatusPal EncryptMessage(SafeDeleteContext securityContext, byte[] input, int offset, int size, int headerSize, int trailerSize, ref byte[] output, out int resultSize)
+        public static unsafe SecurityStatusPal EncryptMessage(SafeDeleteContext securityContext, byte[] input, int offset, int size, int headerSize, int trailerSize, ref byte[] output, out int resultSize)
         {
             // Ensure that there is sufficient space for the message output.
+            int bufferSizeNeeded;
             try
             {
-                int bufferSizeNeeded = checked(size + headerSize + trailerSize);
-
-                if (output == null || output.Length < bufferSizeNeeded)
-                {
-                    output = new byte[bufferSizeNeeded];
-                }
+                bufferSizeNeeded = checked(size + headerSize + trailerSize);
             }
-            catch (Exception e)
+            catch
             {
-                if (!ExceptionCheck.IsFatal(e))
+                if (GlobalLog.IsEnabled)
                 {
-                    if (GlobalLog.IsEnabled)
-                    {
-                        GlobalLog.Assert("SslStreamPal.Windows: SecureChannel#" + LoggingHash.HashString(securityContext) + "::Encrypt", "Arguments out of range.");
-                    }
-
-                    Debug.Fail("SslStreamPal.Windows: SecureChannel#" + LoggingHash.HashString(securityContext) + "::Encrypt", "Arguments out of range.");
+                    GlobalLog.Assert("SslStreamPal.Windows: SecureChannel#" + LoggingHash.HashString(securityContext) + "::Encrypt", "Arguments out of range.");
                 }
-
+                Debug.Fail("SslStreamPal.Windows: SecureChannel#" + LoggingHash.HashString(securityContext) + "::Encrypt", "Arguments out of range.");
                 throw;
             }
-
-            byte[] writeBuffer = output;
+            if (output == null || output.Length < bufferSizeNeeded)
+            {
+                output = new byte[bufferSizeNeeded];
+            }
 
             // Copy the input into the output buffer to prepare for SCHANNEL's expectations
-            Buffer.BlockCopy(input, offset, writeBuffer, headerSize, size);
+            Buffer.BlockCopy(input, offset, output, headerSize, size);
+
+            const int NumSecBuffers = 4; // header + data + trailer + empty
+            var unmanagedBuffer = stackalloc Interop.SspiCli.SecBuffer[NumSecBuffers];
```

Comment:
Maybe?  It's the same name that's used elsewhere for this case, and it does point to the whole buffer which contains several elements.  I think I'll leave it as is for now.